Add Shared Library

### Comment:

The libf2c makefile.u file has a target for libf2c.so. It is not currently built since it is neither part of the "all" target, nor the "install" target.

We should modify the build.sh file to compile this target and add publish the shared library, at least on unix systems.

error:

```
cc -shared -o libf2c.so f77vers.o i77vers.o main.o s_rnge.o abort_.o exit_.o getarg_.o iargc_.o getenv_.o signal_.o s_stop.o s_paus.o system_.o cabs.o ctype.o derf_.o derfc_.o erf_.o erfc_.o sig_die.o uninit.o pow_ci.o pow_dd.o pow_di.o pow_hh.o pow_ii.o pow_ri.o pow_zi.o pow_zz.o c_abs.o c_cos.o c_div.o c_exp.o c_log.o c_sin.o c_sqrt.o z_abs.o z_cos.o z_div.o z_exp.o z_log.o z_sin.o z_sqrt.o r_abs.o r_acos.o r_asin.o r_atan.o r_atn2.o r_cnjg.o r_cos.o r_cosh.o r_dim.o r_exp.o r_imag.o r_int.o r_lg10.o r_log.o r_mod.o r_nint.o r_sign.o r_sin.o r_sinh.o r_sqrt.o r_tan.o r_tanh.o d_abs.o d_acos.o d_asin.o d_atan.o d_atn2.o d_cnjg.o d_cos.o d_cosh.o d_dim.o d_exp.o d_imag.o d_int.o d_lg10.o d_log.o d_mod.o d_nint.o d_prod.o d_sign.o d_sin.o d_sinh.o d_sqrt.o d_tan.o d_tanh.o i_abs.o i_dim.o i_dnnt.o i_indx.o i_len.o i_mod.o i_nint.o i_sign.o lbitbits.o lbitshft.o h_abs.o h_dim.o h_dnnt.o h_indx.o h_len.o h_mod.o h_nint.o h_sign.o l_ge.o l_gt.o l_le.o l_lt.o hl_ge.o hl_gt.o hl_le.o hl_lt.o ef1asc_.o ef1cmc_.o f77_aloc.o s_cat.o s_cmp.o s_copy.o backspac.o close.o dfe.o dolio.o due.o endfile.o err.o fmt.o fmtlib.o ftell_.o iio.o ilnw.o inquire.o lread.o lwrite.o open.o rdfmt.o rewind.o rsfe.o rsli.o rsne.o sfe.o sue.o typesize.o uio.o util.o wref.o wrtfmt.o wsfe.o wsle.o wsne.o xwsne.o dtime_.o etime_.o
Undefined symbols for architecture arm64:
  "_MAIN__", referenced from:
      _main in main.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:89: libf2c.so] Error 1
```

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -10,7 +10,13 @@ patch arithchk.c ${RECIPE_DIR}/patch_arithchk
 # Using the makefile provided with the package
 # but adding the -fPIC option to the CFLAGS
 sed 's/CFLAGS = -O/CFLAGS = -O -fPIC/g' makefile.u > Makefile
-sed -i '' -e "s;\$(CC) -shared;\$(CC) -shared \$(CFLAGS) -Wl,-U,_MAIN__ -Wl,-rpath,${PREFIX}/lib ;g" Makefile
+
+# If this is a mac, allow the main symbol to be undefined in the shared library
+if [ "$(uname)" == "Darwin" ]; then
+  sed -i '' \
+  -e "s;\$(CC) -shared;\$(CC) -shared \$(CFLAGS) -Wl,-U,_MAIN__ -Wl,-rpath,${PREFIX}/lib ;g"\
+  Makefile
+fi
 
 # The Makefile directly calls "cc", which is sometimes not the name
 # of the compiler. So we create a link here and add this directory

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -10,6 +10,7 @@ patch arithchk.c ${RECIPE_DIR}/patch_arithchk
 # Using the makefile provided with the package
 # but adding the -fPIC option to the CFLAGS
 sed 's/CFLAGS = -O/CFLAGS = -O -fPIC/g' makefile.u > Makefile
+sed -i '' -e "s;\$(CC) -shared;\$(CC) -shared \$(CFLAGS) -Wl,-U,_MAIN__ -Wl,-rpath,${PREFIX}/lib ;g" Makefile
 
 # The Makefile directly calls "cc", which is sometimes not the name
 # of the compiler. So we create a link here and add this directory
@@ -20,18 +21,26 @@ export PATH=${PATH}:${PWD}
 make hadd
 make all
 
-# Make sure ${PREFIX}/lib exists, it might not since
+# Make sure ${PREFIX}/ dirs exist, they might not since
 # this package does not depend on anything
 mkdir -p ${PREFIX}/lib
+mkdir -p ${PREFIX}/include/f2c
 
-make install LIBDIR=${PREFIX}/lib 
+# Install the header and static library
+make install libdir=${PREFIX}/lib \
+  LIBDIR=${PREFIX}/lib \
+  includedir=${PREFIX}/include/f2c \
+  INCDIR=${PREFIX}/include/f2c
 
-# Some programs need f2c.h to be in ${PREFIX/include, some in ${PREFIX}/include/f2c
-# "make install" above does not install it, so we need to do it manually
-mkdir -p ${PREFIX}/include
-cp f2c.h ${PREFIX}/include
-mkdir ${PREFIX}/include/f2c
-cp f2c.h ${PREFIX}/include/f2c
+# Some programs need f2c.h to be in ${PREFIX}/include, some in ${PREFIX}/include/f2c
+# "make install" above does not reliably install it, so we need to do it manually.
+cp f2c.h ${PREFIX}/include/f2c.h
+cp f2c.h ${PREFIX}/include/f2c/f2c.h
+
+# Make the shared library explicity. It's neither part of the "all" target, nor the
+# "install" target.
+make libf2c.so
+cp libf2c.so ${PREFIX}/lib/libf2c.so
 
 # Now build the f2c executable
 cd ../src
@@ -46,5 +55,5 @@ cp f2c ${PREFIX}/bin/f2c
 
 # Install the pkg-config file
 mkdir -p ${PREFIX}/lib/pkgconfig
-echo "Version: ${PKG_VERSION}" >> ${RECIPE_DIR}/f2c.pc
 cp ${RECIPE_DIR}/f2c.pc ${PREFIX}/lib/pkgconfig/f2c.pc
+echo "Version: ${PKG_VERSION}" >> ${PREFIX}/lib/pkgconfig/f2c.pc

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -16,21 +16,19 @@ source:
     sha256: {{ srcsha256 }}
 
 build:
-  number: 1005
+  number: 1006
   skip: True  # [vc < 14 and win]
 
 requirements:
   build:
     - {{ compiler('c') }}
     - make
-  host:
-  run:
 
 test:
   commands:
     - test -f ${PREFIX}/include/f2c/f2c.h  # [not win]
     - f2c --version
-    - f2c test_main.f && ${CC} test_main.c -L${PREFIX}/lib -lf2c -I${PREFIX}/include -o test_main && ./test_main  # [not win]
+    - f2c test_main.f && ${CC} test_main.c -L${PREFIX}/lib -lf2c -I${PREFIX}/include -o test_main -Wl,-rpath,${PREFIX}/lib && ./test_main  # [not win]
   requires:
     - {{ compiler('c') }}
   files:

